src/ray/raylet/raylet.cc

```diff
@@ -1,56 +1,129 @@
 #include "raylet.h"
 
+#include <boost/asio.hpp>
 #include <boost/bind.hpp>
+#include <boost/date_time/posix_time/posix_time.hpp>
 #include <iostream>
 
 #include "ray/status.h"
 
 namespace ray {
 
-Raylet::Raylet(boost::asio::io_service &io_service, const std::string &socket_name,
-               const ResourceSet &resource_config,
+namespace raylet {
+
+Raylet::Raylet(boost::asio::io_service &main_service,
+               std::unique_ptr<boost::asio::io_service> object_manager_service,
+               const std::string &socket_name,
+               const NodeManagerConfig &node_manager_config,
                const ObjectManagerConfig &object_manager_config,
-               std::shared_ptr<ray::GcsClient> gcs_client)
-    : acceptor_(io_service, boost::asio::local::stream_protocol::endpoint(socket_name)),
-      socket_(io_service),
-      tcp_acceptor_(io_service,
-                    boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), 0)),
-      tcp_socket_(io_service),
-      object_manager_(io_service, object_manager_config, gcs_client),
-      node_manager_(socket_name, resource_config, object_manager_),
-      gcs_client_(gcs_client) {
-  ClientID client_id = RegisterGcs();
-  object_manager_.SetClientID(client_id);
+               std::shared_ptr<gcs::AsyncGcsClient> gcs_client)
+    : acceptor_(main_service, boost::asio::local::stream_protocol::endpoint(socket_name)),
+      socket_(main_service),
+      object_manager_acceptor_(
+          main_service, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), 0)),
+      object_manager_socket_(main_service),
+      node_manager_acceptor_(
+          main_service, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), 0)),
+      node_manager_socket_(main_service),
+      gcs_client_(gcs_client),
+      object_manager_(main_service, std::move(object_manager_service),
+                      object_manager_config, gcs_client),
+      node_manager_(main_service, node_manager_config, object_manager_, gcs_client_) {
   // Start listening for clients.
   DoAccept();
-  DoAcceptTcp();
+  DoAcceptObjectManager();
+  DoAcceptNodeManager();
+
+  RAY_CHECK_OK(RegisterGcs(main_service, node_manager_config));
+
+  RAY_CHECK_OK(RegisterPeriodicTimer(main_service));
+}
+
+Raylet::~Raylet() {
+  (void)gcs_client_->client_table().Disconnect();
```

Comment:
Is this to avoid a compiler warning? We should not use C style casts. Maybe we can just use the `UNUSED` macro?